### ISSUE TITLE
Gettting NaN for predict.gradient_descent_mse

### ISSUE THREAD

**Issue by won-bae** [google/neural-tangents#128]
Hi,

As the title says, I am getting NaN values for predict.gradient_descent_mse (t=None, diag_reg=0). According to my diagnosis, it is because of cho_factor function in predict._get_cho_solve outputs NaNs, which is weird since cho_factor should not give NaNs for PSD matrix and an empirical tangent kernel (=  ntk_fn(X_train, None, params)) is supposed to be a PSD matrix.
If I give diag_reg > 0, it does not blow up to NaNs but still returns a large number. I may increase diag_reg to avoid this at the cost of accuracy. What would be the best way to get around this problem?

**Comment by won-bae** [comment]
Thanks you for the replies! Unfortunately, as the code is a part of a large pipeline, it is hard for me to provide a reproducible colab notebook. Good thing is that, though, it worked out eventually. Here are what I have tried.

1. using 64-bit float indeed improved the stability but cho_solve still outputs large numbers.
2. `A := (A + A.T) / 2` makes a very small change on `k_train_train` as it is already symmetric.
3. I did not try QR but SVD decomposition indeed helped. I followed https://stackoverflow.com/questions/31251689/how-to-invert-numpy-matrices-using-singular-value-decomposition According to my experience, taking inverse using SVD is much more stable, especially when the dimension of `k_train_train` is large.
4. One crucial things was `jax.jit`. Originally, I wrapped predict.gradient_descent_mse with another function decorated with jit (which I did not mention in the question, sorry). But, if I take the decoration out, I do not get NaNs most of the times (I still get NaNs for further computations).

In conclusion, without `jax.jit` decoration and replacing cholesky with SVD decomposition, my problem was solved. But, I still don't understand why `jax.jit` makes the computation unstable. Do you have any ideas?

**Comment by romanngg** [comment]
Thanks for the update - I don't know myself why JIT would make it unstable, I'd suggest filing a bug with JAX if you manage to provide a repro. In theory XLA optimizations could change the computation slightly / reorder contractions / etc, but in practice I haven't seen it creating problems.

**Comment by won-bae** [comment]
I see. Thank you for the advice. I will try to make a repro and file it with JAX. Closing it as my problem was resolved.
